부산대 BE 정윤성 3단계 - 위시 리스트

dev-yunseong opened the pull request:
##  구현 항목
### 리뷰 반영
  * exception
    * bad request exception과 커스텀 exception을 illegal argument exception으로 통일 했습니다.
    * no_content status에서 bad request status로 수정했습니다.
### 위시 리스트 구현
* db 위시 리스트 table 구현
  * wishes table을 추가했습니다.
    * reference로 product id, member id를 명시적으로 연결했습니다.
    * product id, member id에 unique 제한 사항을 추가했습니다. 
  * 등록된 상품 조회
    * 복잡한 sql 쿼리를 통해 dbms로 filter등 연산을 넘겼습니다.
  * 상품 추가
  * 상품 삭제
* 사용자 정보 받기
  * Authorization 필드의 JWT 토큰 해석하기
    * member service에 있던 jwt 관련 메서드를 JwttokenProvider로 옮겼습니다.
    * jwt token을 파싱해서 member 객체로 만들어주는 메서드를 추가했습니다.
  * 전처리에 연결
    * Header를 읽고 Member로 변환해주는 resolver를 구현했습니다.
    * WebConfig로 resolver를 등록해줬습니다.

* 테스트 코드를 작성했습니다.
  * member 추가 -> wish 추가 -> get wish list 검증 -> wish 수정 -> g
et wish list 검증
  * 다음과 같은 시나리오의 end to end test를 구현했습니다.

## 질문 사항
* 테스트 코드는 어디까지 작성해야할지, 테스트 코드도 모듈화를 해야하는지 궁금합니다!

dev-yunseong committed:
refactor replace custom exceptions and various exceptions with IllegalArgumentException for error handling

dev-yunseong committed:
feature product를 못 찾았을 때 상태 메시지를 badrequest로 수정

dev-yunseong committed:
doc READEME step3 위시 리스트 요구 사항으로 최신화

dev-yunseong committed:
feature schema.sql - wishes table 구현

dev-yunseong committed:
feature JwtTokenProvider - jwt token을 다루는 class로 분리, token을 통해서 Member 객체를 만들어주는 기능 구현

dev-yunseong committed:
feature LoginMemberArgumentResolver -
Header Authentication을 읽고 Member로 변환해주는 Resolver 구현, 등록

dev-yunseong committed:
feature Wish Repository -
* wish product list를 반환하는 메서드 구현
* wish 등록, 삭제, 업데이트하는 메서드 구현
* wish의 count를 가져오는 메서드 구현

dev-yunseong committed:
feature WishService
- wish를 저장, 업데이트하는 메서드 구현
- wish response dto를 반환하는 메서드 구현

dev-yunseong committed:
feature Wish Controller -
* @LoginMember를 통해 Member를 주입 받아서 구현
  * Member id를 Long으로 수정
* wish list를 반환하는 endpoint
* wish를 저장하는 endpoint
* wish count를 업데이트 하는 endpoint

dev-yunseong committed:
refactor Product
- domain이 dto를 알지 못 하도록 구현

dev-yunseong committed:
feature 예외처리 구현
- exception handler 추가
- jwt token parsing에 validation 추가
- wish repository에 예외 던지는 로직 추가

dev-yunseong committed:
feature 테스트 코드 작성
- member 추가 -> wish 추가 -> get wish list 검증 -> wish 수정 -> get wish list 검증
end to end test 구현

Gyuchool commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+
+public class LoginMemberArgumentResolver implements HandlerMethodArgumentResolver {
+
+    private static final String HEADER_KEY = "Authorization";
```
상수로 빼는 습관👍
Authorization처럼 많이 사용하는 헤더는 HttpHeaders.AUTHORIZATION 와 같이 spring에서 제공해주는 클래스가 있으니 이를 활용해도 좋겠네요!

dev-yunseong replied:
반영하겠습니다!

Gyuchool commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Override
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+         String authString = webRequest.getHeader(HEADER_KEY);
```
authString이라는 네이밍만 보면은 어떤 의미인지 이해하기 어려운 것 같습니다! 적합한 다른 네이밍 없을까요?🙄

dev-yunseong replied:
authrizationHeader 어떨까요??
수정해보겠습니다!

Gyuchool commented on src/main/java/gift/controller/MemberController.java:
```diff
@@ -23,7 +23,7 @@ public MemberController(MemberService memberService) {
     public ResponseEntity<AuthResponseDto> registerMember(
             @Validated @RequestBody MemberRequestDto memberRequestDto
-    ) throws BadRequestException {
+    ) {
```
리뷰 반영 감사합니다🙏

Gyuchool commented on src/main/java/gift/controller/WishController.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @GetMapping("")
+    public ResponseEntity<List<WishResponseDto>> getWishList(@LoginMember Member member) {
```
List<>로 바로 반환하는 것보다 객체로 한번 더 감싸서 반환하는 것은 어떨까요? 이렇게하면 어떤 장점 혹은 단점이 있을까요?

dev-yunseong replied:
장점은 어떤 data에 대한 meta 데이터를 추가 하기 쉬워져서 확장성이 좋아질 것 같습니다!
단점은 객체로 감싸는 컴퓨팅 비용이 들고, 약간의 네트워크 코스트가 증가할 것 같습니다

client 측에서도 객체로 파싱해서 다루면 되니깐 응답을 다루기도 편하겠네요!
반영하겠습니다!

Gyuchool replied:
학습을 잘 하셨네요!👍 제가 생각할때도 단점보단 장점이 명확한 것 같네요

Gyuchool commented on src/main/java/gift/repository/WishRepository.java:
```diff
@@ -0,0 +1,99 @@
+            throw new IllegalArgumentException("not found wish");
+    }
+
+    public int getWishCount(long memberId, long productId) {
```
사용하지 않는 메서드인 것 같은데 확인 부탁드립니다!

dev-yunseong replied:
그렇네요! 제거하겠습니다!

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -18,30 +15,30 @@
 
     private final MemberRepository memberRepository;
-    private final String secretKey;
+    private final JwtTokenProvider jwtTokenProvider;
```
Token생성 관련한 책임을 TokenProvider로 넘겼군요👍 
그렇다면 한발 더 나아가서 MemberService가 Jwt에 대한 정보를 알 필요가 있을까요?🤔

dev-yunseong replied:
그럼 Member Service Member를 반환하고 
그 걸 받아서 jwt 를 만들어주는 service를 배치하는게 좋을까요?
controller -> [요청] - >member service
controller <- [member] <- member service
controller -> [member] -> jwt provider
controller <- [jwt token] <- jwt provider

---
아니면 인증을 총괄하는 service가 있어야할까요?
controller -> [요청] - > auth service -> [요청] member service
                                       auth service <- [member] <- member service
                                      auth service -> [member] -> jwt provider 
controller <- [jwt token] <-auth service  <- [jwt token] <- jwt provider

dev-yunseong replied:
위에 방식으로 하면 Controller의 책임이 커지니깐 아래를 선택한다고 한다면
Member Controller가 Authorization Service를 호출해도 되는 건가요??
Authorization Controller로 수정하자니 url이 /api/members 라서 애매한 것 같습니다.

어떻게 하는게 좋을까요?

---
이러한 책임에 대해서 공부하려면 어떻게 해야하나요?

Gyuchool replied:
학습 방식까지 고민하시다니 좋네요!!💯

저도 인증을 처리하는 서비스가 있는게 좋아보이네요..! 회원가입/로그인은 인증 서비스, 유저 관리는 유저 서비스

책임에 대한 학습은 우선 [객체지향의 사실과 오해](https://product.kyobobook.co.kr/detail/S000001628109)라는 책을 읽으면 도움이 될 것 같습니다! 제가 느꼈을떈 약간 추상적이긴 했는데 감각익히기에는 좋습니다:)
이후에는 많이 구현하고 연습하면서 학습했던 것 같네요.!

dev-yunseong replied:
감사합니다!

Gyuchool commented on src/main/java/gift/util/JwtTokenProvider.java:
```diff
@@ -0,0 +1,46 @@
+                .compact();
+    }
+
+    public Member parseJwtToken(String jwtToken) {
```
- jwtToken을 파싱했는데 member를 반환하는게 어색한 것 같아요🤔 
   - 하나의 메서드가 두 가지 일을 하는건 아닌지 의심할 필요가 있겠네요!

dev-yunseong replied:
넵 수정해보겠습니다!

Gyuchool commented on src/test/java/gift/controller/WishControllerTest.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    @Test
+    void Wish_추가_수정_get_list_테스트() {
```
하나의 테스트에 모든 기능을 테스트하는 것보다 각 기능마다 테스트를 추가하는게 좋을 것 같아요! 

현재 제가 테스트했을땐 테스트가 깨지는데 어느 기능에서 문제인지 찾으려면 테스트 코드를 전부 읽어봐야하는 상황이네요!

dev-yunseong replied:
넵 수정했습니다!

Gyuchool reviewed (changes requested):
안녕하세요 윤성님!
마지막 미션도 빠르게 구현해주셨네요! 기능 단위 커밋 및 요구 사항 정리까지 너무 좋은 것 같습니다👍
기능은 정상적으로 동작하는 것을 확인했으나 테스트가 깨지는 곳이 있어서 확인 부탁드릴게요!

추가로 몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!

dev-yunseong committed:
refactor Product 일반 class로 수정
inferStatus를 일반 메서드로 수정

dev-yunseong committed:
refactor parse jwt token의 책임을 나누기
- Jwt Payload를 반환하고
- to member를 통해서 Member로 변환

dev-yunseong committed:
chore get wish count 메서드 삭제

dev-yunseong committed:
feature wishes response dto - list <wish response dto>가 아닌 객체로 리턴하도록 구현

dev-yunseong committed:
refactor Wish Controller Test 테스트 분리

dev-yunseong committed:
feature implement AuthService - member service에서 jwt에 대한 책임 없애기

dev-yunseong commented:
리뷰 감사드립니다!

리뷰 반영을 해봤습니다.
* Product의 inferstatus 메서드의 static을 때도록 수정했습니다.
* parseJwtToken 메서드의 책임을 분배 했습니다.
  * JwtPayload를 반환하도록 수정했습니다. [claimes를 가진 객체]
  * jwtPayload.toMember를 통해 member로 변환할 수 있도록 구현했습니다.
* List<WishResponseDto> 로 응답하던 것을 WishesResponseDto로 수정했습니다.
* Wish Controller Test에서 테스트를 분리시켰습니다.
* Member Service에 JWT에 대한 것을 auth service로 분리시켰습니다.

Gyuchool commented on src/main/java/gift/config/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,38 @@
+    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) throws Exception {
+         String authorizationHeader = webRequest.getHeader(HttpHeaders.AUTHORIZATION);
+
+         if (authorizationHeader == null || !authorizationHeader.startsWith(AUTHORIZATION_PREFIX)) {
```
리뷰 반영 감사합니다😊

Gyuchool commented on src/main/java/gift/util/JwtTokenProvider.java:
```diff
@@ -0,0 +1,39 @@
+        return Jwts.builder()
+                .subject(Long.toString(member.id()))
+                .claim("email", member.email())
+                .signWith(Keys.hmacShaKeyFor(secretKey.getBytes()))
```
`Keys.hmacShaKeyFor(secretKey.getBytes())` 중복되는 메서드가 보이네요!
생성자에서 처리하면 어떨까요?

dev-yunseong replied:
그렇네요! 수정하겠습니다!

Gyuchool commented on src/main/java/gift/util/JwtPayload.java:
```diff
@@ -0,0 +1,19 @@
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.JwtException;
+
+public record JwtPayload(Claims claims) {
```
jwtPayload는 어떤 역할을 위해서 생성하신 걸까요? 
Claims는 무엇이기에 상태로 가지고 있는 것일까요?

dev-yunseong replied:
저번 리뷰에서 parseJwtToken의 책임을 분리하라고 하셨을 때,
jwt를 파싱 했을 때 나와야하는 것은 jwt의 payload라고 생각했습니다!

실제로 파싱 했을 때는 claims가 나오는데, 그것을 wrapper class로 감싸서 관리하는 것이 좋겠다 생각했습니다.

그래서 jwt payload의 역할은 jwt claims의 wrapper class입니다!
Claims는 jwt에 저장 되어있는 정보라 생각했습니다!

claims에서 정보를 아니면 parse jwt token이 id를 반환하여, MemberService에서 Member로 바꾸어 받아오는게 좋을까요?
jwt payload라는 wrapper class를 유지한체 거기서 id를 반환 받아서 Member로 받아오는게 나을까요?

Gyuchool replied:
이전의 리뷰 의도는 jwtTokenProvider가 Member를 반환한 부분이 어색해서 리뷰를 남겼었어요! tokenProvider가 token을 만들거나 값을 꺼내는 역할인데 Member객체까지 만드는게 어색해서요:)
따라서 payload만 꺼내는게 jwtTokenProvider의 책임이 아닐까싶었습니다..!

지금 구조도 좋은 거 같아요👍 토큰에 다른 정보가 추가된다면? 지금 구조가 변화에 대응하긴 좋은 것 같네요:)
다만, util 패키지에 위치한건 어색한 것 같네요🤔 util성 클래스는 어떤 역할을 가진 클래스들일까요?

Gyuchool commented on src/main/java/gift/util/JwtPayload.java:
```diff
@@ -0,0 +1,19 @@
+        }
+
+        Long id = Long.parseLong(idString);
+        return new Member(id, email, "");
```
id의 역할이 무엇이라고 생각하실까요? 토큰에서 꺼낸 id가 저희가 발급한 유저의 id라는 것을 확신할 수 있을까요?🤔

dev-yunseong replied:
id의 역할은 database에서 정보를 찾을 때 사용되는 key인 것 같습니다!

jwt 서명 되어있으니깐 믿어야하는 거 아닌가요??
database에 저장 되어있는지 확인해야할까요??

그렇다고 한다면 공격자가 id를 위조해서 서명까지 해서 요청을 보냈다는 건데, 그럼 auto increament로 id를 지정하는 지금 시스템에선, 1, 2 같은 id로 접근 시도 하면 admin 권한까지 뚫을 수 있을 것 같습니다!

id를 더욱 복잡하게 생성하여서 어떤 id가 있는지 숨겨야할까요??

dev-yunseong replied:
id를 통해서 다른 정보들을 검증하는게 좋을까요??

Gyuchool replied:
좋은 방향으로 고민하셨군요👍
1. 서명은 무엇일까요? (미션과 멀어지는 내용으로 간략하게 자답할게요:)) 
    - 서명은 위조가 안되었다는 것을 보장하는 것이지 토큰 내부 값이 신뢰할 수 있다는 보장은 하지 않습니다(무결성 보장)
    - 따라서 내부 페이로드를 검증하는 부분이 필요합니다:)
    - 참고 자료: [jwt.io](https://jwt.io/)

2. 그러면 제가 생각했을때는 검증하기 위한 두 가지 방법이 떠오르네요🤔
    1. ArgumentResolver에서 db에 저장된 유저인지 확인 후, Member 반환
    2. Payload값만 Controller에 넘기고 Service계층에서 유효성 검증

윤성님은 위 두 가지 선택지중 어느 방식이 적합하다고 생각하시는지 장단점을 비교해서 의견 공유해주세요!! 혹은 다른 방법이 떠오르시다면 말씀주셔도 좋아요!

id를 복잡하게 생성하는 것도 좋은 의견이에요! 다만 토큰에 모든 유저 정보를 담을 수 없으니 현재처럼 Member객체를 만들어서 넘기는 것은 불가능할 것 같네요🤔

dev-yunseong replied:
뭔가 payload 상태로 데이터가 다니는 것 보다 resolver에서 db를 보고 Member를 반환하는게 좋을 것 같습니다!

Gyuchool commented on src/test/java/gift/controller/WishControllerTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    @Test
+    void Wish_추가_테스트() {
+        String token = registerMemberAndGetToken("demo1@email", "asdf");
```
E2E테스트(인수테스트)의 목적은 무엇이라고 생각하실까요? 의견이 궁금합니다!

dev-yunseong replied:
시스템을 사용자가 사용하는 것을 모방하여 전체적인 시스템을 테스트하는 것인 것 같습니다!

Gyuchool replied:
참고 자료를 읽어보시면 학습하는데 도움될 것입니다👍

- https://tecoble.techcourse.co.kr/post/2021-05-25-unit-test-vs-integration-test-vs-acceptance-test/
- https://brunch.co.kr/@boorownie/6

dev-yunseong replied:
감사합니다~

Gyuchool commented on src/test/java/gift/controller/WishControllerTest.java:
```diff
@@ -0,0 +1,99 @@
+
+        ResponseEntity<WishesResponseDto> responseEntity = getWishList(token);
+
+        assertThat(responseEntity.getBody().wishes()).isNotNull();
+        assertThat(responseEntity.getBody().wishes().size()).isGreaterThan(0);
+        assertThat(responseEntity.getBody().wishes().getLast().count()).isEqualTo(123);
+        assertThat(responseEntity.getBody().wishes().getLast().productResponseDto().id()).isEqualTo(1);
```
```suggestion
        assertAll(
            () -> assertThat(responseEntity.getBody().wishes()).isNotNull(),
            () -> assertThat(responseEntity.getBody().wishes().size()).isGreaterThan(0),
            () -> assertThat(responseEntity.getBody().wishes().getLast().count()).isEqualTo(123),
            () -> assertThat(responseEntity.getBody().wishes().getLast().productResponseDto().id()).isEqualTo(1)       
        )
```

이렇게 assertAll()로 묶으면 어느 장단점이 있을까요?

dev-yunseong replied:
장점은 앞에 assert가 실패해도 뒤 assert까지 검사할 수 있는 것 같습니다.
단점은 코드가 비대해지는 것 같습니다!

이렇게 수정해보겠습니다

Gyuchool commented on src/main/java/gift/service/WishService.java:
```diff
@@ -0,0 +1,50 @@
+        if (count == 0)
+            return;
+
+        wishRepository.saveWish(memberId, productId, count);
```
```suggestion
        if (count == 0){
            return;
        }

        wishRepository.saveWish(memberId, productId, count);
```
if()문을 쓸때는 중괄호를 써주는게 가독성 측면에서 더 좋을 것 같아요!
save시도하기 전에 검증해서 미리 리턴해준게 성능적으로 좋네요👍

dev-yunseong replied:
넵 알겠습니다1

Gyuchool commented on src/main/java/gift/service/AuthService.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    public AuthResponseDto register(MemberRequestDto memberRequestDto) {
+        return new AuthResponseDto(
+                jwtTokenProvider.makeJwtToken(
+                        memberService.creatMember(memberRequestDto)
+                ));
```
중첩된 호출은 가독성 및 디버깅이 어려울 것 같은데 아래처럼 리팩토링하는 것은 어떠신가요? 
```suggestion
        Member member = memberService.createMember(memberRequestDto);
        String token = jwtTokenProvider.makeJwtToken(member);
        return new AuthResponseDto(token);
```

dev-yunseong replied:
넵 수정하겠습니다!

Gyuchool commented on src/main/java/gift/service/AuthService.java:
```diff
@@ -0,0 +1,33 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class AuthService {
```
책임에 맞게 인증 서비스 분리 좋네요👍

Gyuchool commented on src/main/java/gift/service/ProductService.java:
```diff
@@ -37,14 +32,18 @@ public void deleteProductById(Long id) {
     public List<ProductResponseDto> findApprovedProducts() {
         return productRepository.findAllProducts().stream()
-                .filter(product -> product.status() == Product.Status.APPROVED)
+                .filter(product -> product.getStatus() == Product.Status.APPROVED)
```
객체지향 생활 체조 원칙! getter 사용을 지양하자!

왜 getter사용을 지양하라고 하는 것일까요? 어떻게 개선할 수 있을까요?

dev-yunseong replied:
capsulization이 흐릿해져서 그런 것일까요??
로직을 객체 안으로 넣으면 될 것 같습니다!

Gyuchool reviewed (changes requested):
안녕하세요 윤성님!
역시나 빠르게 리뷰 반영 및 구현을 완료하셨네요..!! 대단합니다👍💯
다만, 테스트가 실패하는 부분이 있는데 확인 부탁드립니다🙏

몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 다시 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!

dev-yunseong committed:
refactor secretKey를 객체 형태로 저장하도록 수정

dev-yunseong committed:
feature JwtPayload - claim들 변환해서 id만 저장하도록 수정, member로 변환하는 책임은 없앰

dev-yunseong committed:
refactor JwtToken
- memberId만 사용되기에 그것만 jwt에 저장하도록 구현

dev-yunseong committed:
refactor Member Resolver -
member id를 가지고 member service에서 member를 가져오도록 수정

dev-yunseong committed:
feature Member - id를 가지고 Member를 가지고 오는 기능 구현

dev-yunseong committed:
refactor Product - approve 된 것을 확인 하는 로직은 메서드로 수정

dev-yunseong committed:
refactor if 문에 중괄호 명확히 표시

dev-yunseong committed:
fix jwt payload validation - email을 저장하지 않는 것으로 수정한 것 반영

dev-yunseong committed:
refactor wish controller test assert all 적용

dev-yunseong commented:
리뷰 감사드립니다!
오늘도 한번 리뷰 반영해봤습니다.

그 제 로컬 환경에서는 테스트가 실패하지 않는데 혹시 어디에서 실패가 나는 걸까요??

## 수정사항
* test에서 assert all을 적용해봤습니다.
* product 객체 안에서 approve 확인을 하도록 수정했습니다. [getter 지양]
* member를 id를 가지고 가지고 오도록, service, repository에 메서드를 추가했습니다.
* jwt payload에서 member로 변환하는 로직을 빼고, id만 반환하도록 수정했습니다. -> 그리고 resolver에서 member service에서 member를 받아오도록해서 member에 대한 책임을 member service에 넘겼습니다.
* jwt token provider에서 sercet key를 객체 형태로 가지고 있도록 수정했습니다.

Gyuchool commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -18,54 +12,47 @@
+    }
+
+    public Member creatMember(MemberRequestDto memberRequestDto) {
         String passwordHash = BCrypt.hashpw(
```
시간도 남았으니 한발 더 나아가보죠!

Password생성을 Service레이어에서 작성한 것 보다 더 응집도있게 작성하려면 어떻게할까요? 누구의 책임으로 넘기는게 좋을까요?

dev-yunseong replied:
domain으로 넘기면 좋을 것 같습니다!
한번 이쁘게 넘겨 보겠습니다!

Gyuchool reviewed (changes requested):
안녕하세요 윤성님:)
역시나 이번에도 빠르게 학습, 반영 및 의견까지 주셨네요💯💯💯

제 로컬에서 돌렸을땐 아래처럼 테스트가 깨져서 확인 부탁드릴게요! 
<img width="1333" height="523" alt="스크린샷 2025-07-10 오후 11 04 49" src="https://github.com/user-attachments/assets/d17f43f6-6b37-4c07-b7b8-3a91563c91af" />


추가로 몇 가지 코멘트를 남겨두었으니 반영 및 의견 주시고 요청주세요!🙏
궁금한 점은 디스코드 DM이나 PR 커멘트로 언제든지 여쭤봐주세요!

dev-yunseong committed:
refactor Member password, passwordHash 네이밍 구분

dev-yunseong committed:
refactor Member
- 일반 class로 수정
- validation, hashing 로직을 domain으로 옮김
- 생성자 나누기
- password, passwordHash 네이밍 구분

dev-yunseong committed:
fix ProductController Test
- 여러 예상되는 키워드 중 하나만 맞으면 되도록 구현 [local 언어 설정에 따라 다른 듯합니다]

dev-yunseong commented:
리뷰 반영, 수정했습니다!
- product controller test에서 테스트가 깨지던 이슈를 수정했습니다.
  - local 언어 설정에 따라 다른 exception message 때문에 다른 결과가 나왔기 때문에 any match를 통해서 검사하도록 구현했습니다.
- password 평문이랑 password hash 네이밍이 헷갈려서 구분했습니다.
- password hashing의 책임을 Member로 넘겼습니다.

dev-yunseong committed:
chore Jwt 관련 class package gift.security.~로 옮김

Gyuchool commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,4 +1,39 @@
+                BCrypt.gensalt());
+    }
+
+    public Member(Long id, String email, String passwordHash) {
```
다른 개발자가 해당 생성자를 사용해서 member를 생성한다면 평문으로 저장될 수도 있겠네요?🤔 어떻게 방지할 수 있을까요?

dev-yunseong replied:
좀 더 hash를 받는 다거나 평문을 받는지를 명확히 하는 네이밍의 create static 메서드로 받는다면 해결 될까요??
원천 차단하는 방법은 떠오르지가 않네요..

Gyuchool replied:
정적팩토리 메서드 사용, hashPassword 객체로 감싸는 등 다양한 방법이 있을 것 같아요!

Gyuchool commented on src/main/java/gift/entity/Member.java:
```diff
@@ -1,4 +1,39 @@
 package gift.entity;
 
-public record Member(long id, String email, String password) {
+import org.mindrot.jbcrypt.BCrypt;
```
지금도 훌륭하지만 한발 더 나아가볼까요?

member가 암호화를 어떻게 하는지 알 필요가 있을까요?🙄

dev-yunseong replied:
그럼 Member에 대한 Factory class를 구현하고 그곳에서 암호화한 password를 주입해주는게 나을까요??

dev-yunseong replied:
그럼 생성할 때 hashing을 해주는 것은 해결이 되는데
평문 password랑 비교할 때는 어떻게 해야할지 모르겠습니다..
member.validatePassword(String password) 말고 다른 
생성과 password 확인 둘 다 깔끔하게 책임을 나누려면 어떻게 하는게 좋을까요?

Gyuchool replied:
Password객체를 만들고 해당 객체에 책임을 넘기면 좋을 것 같아서 의견드렸습니다:)

Gyuchool commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,26 +1,65 @@
+    }
+
+    public Product(String name, int price, String imageUrl) {
+        this(-1L, name, price, imageUrl);
```
생성자 체이닝 활용👍

Gyuchool commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,26 +1,65 @@
+        this.status = status;
+    }
+
+    public Product(Long id, String name, int price, String imageUrl) {
```
여기도 부생성자를 활용할 수 있겠네요..!
생성자 체이닝을 사용하면 어떤 장점이 있을까요?

dev-yunseong replied:
코드 중복을 줄이고, 필드가 수정되었을 때 코드 수정을 최소한 할 수 있을 것 같습니다!

dev-yunseong replied:
그런데 여기에선 
생성자의 this() 이 인자 안에서 메서드를 호출 못 해서 생성자 체이닝을 활용하지 못 했던 것 같습니다!
```
    public Product(String name, int price, String imageUrl) {
        this(-1L, name, price, imageUrl);
    }

    public Product(Long id, String name, int price, String imageUrl, Status status) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.imageUrl = imageUrl;
        this.status = status;
    }

    public Product(Long id, String name, int price, String imageUrl) {
        Status status = inferStatus(name);
        this(id, name, price, imageUrl, status);
    }
```

다른 방법이 있을까요?? 아님 이건 그대로 둬야할까요?

Gyuchool replied:
정적팩토리 메서드를 고려해도 좋겠네요!

Gyuchool reviewed (approved):
안녕하세요 윤성님!
마지막 단계 미션 진행하시느라 고생하셨습니다!! 미션 중에 많은 대화를 주고 받을 수 있어서 즐거웠습니다:) 👍

코멘트 남겨두었으니 주말동안 고민해보시고 다음 주차 미션도 지금처럼 화이팅하세요💪💪
추가로 코멘트에 대해 궁금한 점이 생기면 언제든 DM주세요👍 
2주간 고생 많으셨습니다🙇‍♂️

dev-yunseong commented:
2주간 리뷰 정말 감사드립니다!
